### PR TITLE
add 'cases' helper to retrieve all instances of the ENUM

### PR DESCRIPTION
**Pull request by timacdonald**
Hey! Thanks so much for the package.

I am currently using this package and have the need to be able to pull out all values from each ENUM.

Here is a silly example of why I want some thing like this...

Imagine you have some ENUMs that represent different types of drinks.

```php
/**
 * @method static self coke()
 * @method static self fanta()
 * @method static self moutainDew()
 */
class Soda extends Enum
{
    //
}

/**
 * @method static self lime()
 * @method static self lemon()
 * @method static self blueberry()
 */
class Cordial extends Enum
{
    //
}

/**
 * @method static self stout()
 * @method static self brown()
 * @method static self red()
 */
class Beer extends Enum
{
    //
}
```

Now within specific environments we want certain values available. This example is within Laravel, but the concept applies to any framework...

```php
// ServiceProvider
public function register()
{
    $this->app->bind(DrinkCollection::class, function (Application $app) {
        // All soda is always available
        return DrinkCollection::make(Soda::all())

            // All Cordials are available in production
            ->when($app->isProduction(), function (DrinkCollection $drinks) {
                return $drinks->concat(Cordial::all());

            // in production 'dark' and 'red' beer is available....
            })->when($app->isProduction(), function (DrinkCollection $drinks) {
                return $drinks->merge([
                    Beer::dark(),
                    Beer::red(),
                ]);

            // outside of production only 'brown' beer is available
            }, function (DrinkCollection $drinks) {
                return $drinks->merge([
                    Beer::brown(),
                ]);
            });
    });
}

// Controller

public function whatever(DrinkCollection $drinks)
{
    // $drinks now contains all available Drink ENUMs for this environment
}
```

I could of course explicitly add each individual Soda and Cordial, but we know we always want _all_ of them, so this little helper allows us to do just that. There is, as you know, a lot of trickery happening under the hood of the package to make it work nicely, so I'm hoping I've gone about this implementation the right way.

Welcoming any feedback to make it better.

### CODE REVIEW

**Comment by Gummibeer** [comment]
Hey,
thanks for this PR. The method logic was already requested by another one - so it seems like we could/should implement it.

But this method name will not get into the package as it is too likely that it will conflict. #75 
The "old" `getAll()` would be an option but I don't like it as it doesn't fit the current naming of the other three similar methods `toArray()`, `toValues()` and `toLabels()`.
So I would prefer a `toXyz()` method here.
The first that comes to mind would be `toInstances()` but I'm not 100% happy with it.

@brendt any genius ideas?^^
#69 was the discussion about the `toValues()` and `toLabels()` methods.

File: src/Enum.php [hunk 1]
```diff
@@ -27,6 +27,16 @@ abstract class Enum implements JsonSerializable
     /** @psalm-var array<string, array<string, \Spatie\Enum\EnumDefinition>>  */
     private static array $definitionCache = [];
 
+    /**
+     * @return static[]
+     */
+    public static function all(): array
+    {
+        $instances = array_map(fn (EnumDefinition $definition) => static::make($definition->value), static::resolveDefinition());
```
Comment (Gummibeer) [inline src/Enum.php:35]:
```suggestion
        $instances = array_map(
            fn (EnumDefinition $definition): self => static::make($definition->value), 
            static::resolveDefinition()
        );
```

File: tests/EnumToArrayTest.php [hunk 2]
```diff
@@ -44,6 +44,15 @@ public function test_to_array_with_value_and_label_map()
             EnumAsArray::toLabels()
         );
     }
+
+    /** @test **/
```
Comment (Gummibeer) [inline tests/EnumToArrayTest.php:48]:
```suggestion
    /** @test */
```

File: src/Enum.php (hunk 1 above)
Comment (timacdonald) [reply src/Enum.php:35]: do you want to return `self` when it in fact returns `static`? Maybe instead it should be `: Enum` instead to match the return type of `static::make()`?

File: tests/EnumToArrayTest.php (hunk 2 above)
Comment (timacdonald) [reply tests/EnumToArrayTest.php:48]: Thanks!

File: src/Enum.php (hunk 1 above)
Comment (Gummibeer) [reply src/Enum.php:35]: Good point. `static` is a PHP8 return type, but if we've used `Enum` in the other methods we should stick with it! 👌

**Comment by timacdonald** [comment]
So I've updated this with the suggestions, however I'm actually going to close this. Re-looking at this I was kinda just abusing an ENUM and ended up going down a different path instead. Of course, the PR is here if you still value in this being included.

**Comment by timacdonald** [comment]
p.s. thanks for the great package(s)

**Comment by Gummibeer** [comment]
Thanks @timacdonald ! 🙏
I think that we will add something like that. But could take a while.

@brendt what do you think?
Should we add this logic? If so which method name? I have `toInstances()` or `toList()` in my mind but don't really like both. 🤔

**Comment by brendt** [comment]
We want to get a list of enum objects, is that right? I just don't want to end up with four methods doing slightly different things… `toArray`, `toValues`, `toLabels`, `toList`. I'm fine adding a way to get a list of instantiated enum objects, but I'm just wondering what the best approach is.

**Comment by Gummibeer** [comment]
That's right. The really missing part is the name. 🤔

**Comment by timacdonald** [comment]
I'm more than happy to make any changes to this PR to help contribute and give back to you. I'll re-open as we seem to be moving towards a potentially wanting the feature.

**Comment by brendt** [comment]
I'm all for it, if we can settle on a name. I actually like `all` the most come to think of it. It isn't prefixed by `to`, which implies some kind of transformation. `all` simply means "all enums". What do you think?

**Comment by Gummibeer** [comment]
Yeah, but `all` is very likely breaking user values.
We already had user code in the issues with an `all` value.

**Comment by brendt** [comment]
😕  any other ideas?

**Comment by Gummibeer** [comment]
`list` is a common name in other languages, but most times via helper classes.
So for us it would be `toList()` or `asList()`.
But most other languages only have a `values` property/method as they don't have labels.

@brendt is there a similar method suggestion in the PHP8.1 enum RFC? 🤔

**Comment by brendt** [comment]
I believe it's called `Enum::cases` because they use the strange `case` notation.

**Comment by timacdonald** [comment]
If we match that naming, do we have the possibility of conflicting if it becomes a reserved word (I haven't been keeping up with the RFC)?

Or is that not a concern considering the RFC is adding Enum's so there is a possibility this package may be "finished" at that point?

**Comment by brendt** [comment]
The package will definitely retire if and when enums land in PHP. I don't think we'll manage to find an idea that works better than `all` (which technically wouldn't be a breaking change btw: PHP does exactly the same by reserving more keywords in a minor release). 

So either we just accept that some people might have to deal with not using `all` anymore as an enum value, or we settle for something less good. I don't really mind `cases`, but let's keep in mind the RFC might still change.

**Comment by timacdonald** [comment]
I guess `allCases` is an option. I prefer `all`...but it is an option.

**Comment by timacdonald** [comment]
I know this is left for dead when ENUM's hit the big time - but in the meantime it would be nice if this was available IMO. Updated the naming to match the final PHP enum naming `Enum::cases()`

**Review by Gummibeer — APPROVED** [review]

**Comment by Gummibeer** [comment]
Thanks for your work @timacdonald ! 🎉 🚀

**Comment by Gummibeer** [comment]
https://github.com/spatie/enum/releases/tag/3.8.0

**Pull request merged**